lib/util/github/graphql/datasource-fetcher.ts

```diff
@@ -266,18 +271,49 @@ export class GithubGraphqlDatasourceFetcher<
       }
     }
 
-    return this.cacheStrategy().finalize();
+    if (this.isPersistent) {
+      await this.storePersistenceFlag(30);
+    }
+  }
+
+  private async doCachedQuery(): Promise<ResultItem[]> {
+    await this.loadPersistenceFlag();
+    if (!this.isPersistent) {
+      await this.doPaginatedFetch();
+    }
+
+    const res = await this.cacheStrategy().finalizeAndReturn();
+    if (res.length) {
+      return res;
+    }
+
+    delete this.isPersistent;
+    await this.doPaginatedFetch();
+    return this.cacheStrategy().finalizeAndReturn();
+  }
+
+  async loadPersistenceFlag(): Promise<void> {
+    const ns = this.getCacheNs();
+    const key = `${this.getCacheKey()}:is-persistent`;
+    this.isPersistent = await packageCache.get<true>(ns, key);
+  }
+
+  async storePersistenceFlag(minutes: number): Promise<void> {
+    const ns = this.getCacheNs();
+    const key = `${this.getCacheKey()}:is-persistent`;
+    await packageCache.set(ns, key, true, minutes);
   }
 
   /**
-   * This method intentionally was made not async, though it returns `Promise`.
-   * This method doesn't make pages to be fetched concurrently.
-   * Instead, it ensures that same package release is not fetched twice.
+   * This method caches results for the same packages during repository run.
+   *
+   * It intentionally was made not async, though it returns `Promise`,
+   * to ensure the same package is prevented from being fetched twice.
```

Comment:
```suggestion
   * This method is intentionally not async, even though it returns a `Promise`.
   * This way Renovate fetches the same package once, instead of multiple times.
```

How about something like this? Please check for techinical accuracy before applying my suggestion! 😉